clean_data error on real ensemble data (cop_mediation)

@benkeser 

make[1]: Entering directory '/trials/covpn/p3003/analysis/correlates_reporting/cop_mediation'
Rscript code/clean_data.R
Warning message:
Project requested R version '4.1.0' but '4.0.4' is currently being used

Attaching package: ‘dplyr’

The following objects are masked from ‘package:stats’:

    filter, lag

The following objects are masked from ‘package:base’:

    intersect, setdiff, setequal, union


Attaching package: ‘kyotil’

The following object is masked from ‘package:dplyr’:

    last

Warning messages:
1: Project requested R version '4.1.0' but '4.0.4' is currently being used
2: In max(full_data[full_data$EventIndPrimaryD57 == 1 & full_data$Trt ==  :
  no non-missing arguments to max; returning -Inf
[1] "Day29"
Rscript code/run_analysis.R
Warning message:
Project requested R version '4.1.0' but '4.0.4' is currently being used

Attaching package: ‘dplyr’

The following objects are masked from ‘package:stats’:

    filter, lag

The following objects are masked from ‘package:base’:

    intersect, setdiff, setequal, union


Attaching package: ‘kyotil’

The following object is masked from ‘package:dplyr’:

    last

here() starts at /trials/covpn/p3003/analysis/correlates_reporting/cop_mediation
Loading required package: nnls
Loading required package: gam
Loading required package: splines
Loading required package: foreach
Loaded gam 1.20

Super Learner
Version: 2.0-28
Package created on 2021-05-04

Warning messages:
1: Project requested R version '4.1.0' but '4.0.4' is currently being used
2: In max(full_data[full_data$EventIndPrimaryD57 == 1 & full_data$Trt ==  :
  no non-missing arguments to max; returning -Inf
[1] "Day29bindSpike"
Loading required package: nloptr
Loading required namespace: glmnet
Loading required namespace: xgboost
Loading required namespace: ranger
Error in SuperLearner::SuperLearner(Y = Y[R == 1 & C == 1], X = data.frame(A = A,  :
  missing data is currently not supported. Check Y, X, and newX for missing values
Calls: natmed2 -> <Anonymous>
In addition: There were 50 or more warnings (use warnings() to see the first 50)
Execution halted
Makefile:16: recipe for target 'output/full_result.rds' failed
make[1]: *** [output/full_result.rds] Error 1
make[1]: Target 'all' not remade because of errors.
make[1]: Leaving directory '/trials/covpn/p3003/analysis/correlates_reporting/cop_mediation'
Makefile:40: recipe for target 'cop_analysis' failed
make: *** [cop_analysis] Error 2

--- a/cop_mediation/code/clean_data.R
+++ b/cop_mediation/code/clean_data.R
@@ -13,85 +13,87 @@ full_data <- read.csv(here::here("..", "data_clean", data_name_updated))
 #   adding 1e-6 to the first cut point helps avoid an error when 33% is the minimial value
 #   -Inf and Inf are added to q.a because otherwise cut2 may assign the rows with the minimum value NA
 
-marker.cutpoints <- list()    
-for (a in assays) {
-    marker.cutpoints[[a]] <- list()    
-    for (ind.t in times) {
-        if(ind.t == "Day57"){
-          dat.vacc.pop=subset(full_data, Trt==1 & Bserostatus==0 & !is.na(wt.D57))
-          wt <- dat.vacc.pop$wt.D57
-        }else{
-          dat.vacc.pop=subset(full_data, Trt==1 & Bserostatus==0 & !is.na(wt.D29))
-          wt <- dat.vacc.pop$wt.D29
-        }
-        q.a <- Hmisc::wtd.quantile(dat.vacc.pop[[ind.t %.% a]], weights = dat.vacc.pop$wt, probs = c(1/3, 2/3))
-        q.a[1]=q.a[1]+1e-6
-        full_data[[ind.t %.% a %.% "cat"]] <- as.numeric(factor(Hmisc::cut2(full_data[[ind.t %.% a]], cuts = c(-Inf, q.a, Inf))))
-        stopifnot(length(table(full_data[[ind.t %.% a %.% "cat"]])) == 3)
-        marker.cutpoints[[a]][[ind.t]] <- q.a
-    }
-}
-
-# Generate the outcome and censoring indicator variables
-for (time in times) {
-  print(time)
-  data <- full_data
-  outcome <-
-    data[[Event_Ind_variable[[time]]]] == 1 & data[[Event_Time_variable[[time]]]] <= tf[[time]]
-  outcome <- as.numeric(outcome)
-
-  if(FALSE) {
-  Delta <-
-    1 - (data[[Event_Ind_variable[[time]]]] == 0 &
-      data[[Event_Time_variable[[time]]]] < tf[[time]])
-  } else {
-    Delta <- rep(1, length(outcome))
+if(!is.null(assays)){
+  marker.cutpoints <- list()    
+  for (a in assays) {
+      marker.cutpoints[[a]] <- list()    
+      for (ind.t in times) {
+          if(ind.t == "Day57"){
+            dat.vacc.pop=subset(full_data, Trt==1 & Bserostatus==0 & !is.na(wt.D57))
+            wt <- dat.vacc.pop$wt.D57
+          }else{
+            dat.vacc.pop=subset(full_data, Trt==1 & Bserostatus==0 & !is.na(wt.D29))
+            wt <- dat.vacc.pop$wt.D29
+          }
+          q.a <- Hmisc::wtd.quantile(dat.vacc.pop[[ind.t %.% a]], weights = dat.vacc.pop$wt, probs = c(1/3, 2/3))
+          q.a[1]=q.a[1]+1e-6
+          full_data[[ind.t %.% a %.% "cat"]] <- as.numeric(factor(Hmisc::cut2(full_data[[ind.t %.% a]], cuts = c(-Inf, q.a, Inf))))
+          stopifnot(length(table(full_data[[ind.t %.% a %.% "cat"]])) == 3)
+          marker.cutpoints[[a]][[ind.t]] <- q.a
+      }
   }
-  Delta <- as.numeric(Delta)
-  data$outcome <- outcome
-  data$Delta <- Delta
+}
+# Generate the outcome and censoring indicator variables
+if(!is.null(times)){
+  for (time in times) {
+    print(time)
+    data <- full_data
+    outcome <-
+      data[[Event_Ind_variable[[time]]]] == 1 & data[[Event_Time_variable[[time]]]] <= tf[[time]]
+    outcome <- as.numeric(outcome)
 
-  data$TwophasesampInd <- as.numeric(data[[twophaseind_variable[[time]]]])
-  data$wt <- data[[weight_variable[[time]]]]
-  #data$grp <- data[[twophasegroup_variable[[time]]]]
-  
-  # Subset data
-  variables_to_keep <-
-    c(
-      covariates,
-      c(outer(times, assays, FUN=paste0)),
-      "TwophasesampInd",
-      #"grp",
-      "Trt",
-      "wt",
-      "outcome",
-      "Delta"
+    if(FALSE) {
+    Delta <-
+      1 - (data[[Event_Ind_variable[[time]]]] == 0 &
+        data[[Event_Time_variable[[time]]]] < tf[[time]])
+    } else {
+      Delta <- rep(1, length(outcome))
+    }
+    Delta <- as.numeric(Delta)
+    data$outcome <- outcome
+    data$Delta <- Delta
+
+    data$TwophasesampInd <- as.numeric(data[[twophaseind_variable[[time]]]])
+    data$wt <- data[[weight_variable[[time]]]]
+    #data$grp <- data[[twophasegroup_variable[[time]]]]
+    
+    # Subset data
+    variables_to_keep <-
+      c(
+        covariates,
+        c(outer(times, assays, FUN=paste0)),
+        "TwophasesampInd",
+        #"grp",
+        "Trt",
+        "wt",
+        "outcome",
+        "Delta"
+      )
+
+    keep <- data$Bserostatus == 0 & !is.na(data$wt)
+    
+    data_keep <- data[keep, variables_to_keep]
+
+    saveRDS(data_keep,
+      file = here::here("data_clean", paste0("data_", time, ".rds"))
     )
 
-  keep <- data$Bserostatus == 0 & !is.na(data$wt)
-  
-  data_keep <- data[keep, variables_to_keep]
+    # swap out categorical markers
+    variables_to_keep_cat <-
+      c(
+        covariates,
+        paste0(c(outer(times, assays, FUN=paste0)), "cat"),
+        "TwophasesampInd",
+        #"grp",
+        "wt",
+        "outcome",
+        "Delta",
+        "Trt"
+      )
+    data_keep_cat <- data[keep, variables_to_keep_cat]
 
-  saveRDS(data_keep,
-    file = here::here("data_clean", paste0("data_", time, ".rds"))
-  )
-
-  # swap out categorical markers
-  variables_to_keep_cat <-
-    c(
-      covariates,
-      paste0(c(outer(times, assays, FUN=paste0)), "cat"),
-      "TwophasesampInd",
-      #"grp",
-      "wt",
-      "outcome",
-      "Delta",
-      "Trt"
+    saveRDS(data_keep_cat,
+      file = here::here("data_clean", paste0("data_", time, "_cat.rds"))
     )
-  data_keep_cat <- data[keep, variables_to_keep_cat]
-
-  saveRDS(data_keep_cat,
-    file = here::here("data_clean", paste0("data_", time, "_cat.rds"))
-  )
- 
+  }
 }

--- a/cop_mediation/code/params.R
+++ b/cop_mediation/code/params.R
@@ -18,8 +18,12 @@ tf <- list("Day57" = tf_Day57, "Day29" = tf_Day29)
 # tf should be large enough that most events are observed but small enough so that not many people are right censored. For the practice dataset, tf = 170 works.
 # Right-censoring is taken into account for  this analysis.
 
+
+# because the SAP specifies that these results depend on blinded evaluation
+# of the overlap assumption, by default the analysis does nothing.
+# As more immuno reports become available, will add in code here.
 if(study_name_code != "COVE"){
-  times <- intersect(c("Day57", "Day29"), times)
+  # times <- intersect(c("Day57", "Day29"), times)
 }else{
   times <- "Day29"
   assays <- c("pseudoneutid50", "pseudoneutid80")

--- a/cop_mediation/code/params.R
+++ b/cop_mediation/code/params.R
@@ -22,7 +22,7 @@ if(study_name_code != "COVE"){
   times <- intersect(c("Day57", "Day29"), times)
 }else{
   times <- "Day29"
-  assays <- "pseudoneutid80"
+  assays <- c("pseudoneutid50", "pseudoneutid80")
 }
 # Marker variables to generate results for
 # assays <- c("bindSpike", "bindRBD", "pseudoneutid80", "liveneutid80", "pseudoneutid80", "liveneutid80", "pseudoneutid50", "liveneutid50")

--- a/cop_mediation/code/run_analysis.R
+++ b/cop_mediation/code/run_analysis.R
@@ -32,89 +32,91 @@ if(fast_run){
 	)
 }
 
-cat_result <- NULL
-quant_result <- NULL
-assay_col <- day_col <- NULL
-for (marker in c(outer(times, assays, FUN=paste0))) {
-  print(marker)
+if(!is.null(times) & !is.null(assays)){
+	cat_result <- NULL
+	quant_result <- NULL
+	assay_col <- day_col <- NULL
+	for (marker in c(outer(times, assays, FUN=paste0))) {
+	  print(marker)
 
-  time <- marker_to_time[[marker]]
+	  time <- marker_to_time[[marker]]
 
-  # quantitative marker
-  data_full <- readRDS(here("data_clean", paste0("data_", time, ".rds")))
-  # # check whether to perform the analysis with quantitative marker
-  # perform_continuous_analysis <- 
-  #   min(data_full[data_full$Trt == 1 , marker], na.rm = TRUE) < 
-  #     max(data_full[data_full$Trt == 0, marker], na.rm = TRUE)
-  perform_continuous_analysis <- TRUE
-  
-  if(perform_continuous_analysis){
-	  fit <- natmed2(
-	    W = data_full[, covariates, drop = FALSE],
-	    A = data_full$Trt,
-	    R = data_full$TwophasesampInd,
-	    S = data_full[, marker, drop = TRUE],
-	    C = data_full$Delta,
-	    Y = data_full$outcome,
-	    gRn = 1 / data_full$wt,
-	    glm_gA = "1",
-	    glm_gAS = NULL,
-      	    SL_gAS = sl_library,
-            glm_QY_WAS = NULL,
-            SL_QY_WAS = sl_library,
-            glm_QY_WACY = NULL,
-            SL_QY_WACY = sl_library,
-            glm_QD_WACY = NULL,
-            SL_QD_WACY = sl_library,
-            glm_QY_W = NULL,
-            SL_QY_W = sl_library,
-            glm_QY_WA = NULL,
-            SL_QY_WA = sl_library
-	  )
-	  this_row <- format_row(fit)
-	}else{
-		this_row <- c(NA, NA, NA)
+	  # quantitative marker
+	  data_full <- readRDS(here("data_clean", paste0("data_", time, ".rds")))
+	  # # check whether to perform the analysis with quantitative marker
+	  # perform_continuous_analysis <- 
+	  #   min(data_full[data_full$Trt == 1 , marker], na.rm = TRUE) < 
+	  #     max(data_full[data_full$Trt == 0, marker], na.rm = TRUE)
+	  perform_continuous_analysis <- TRUE
+	  
+	  if(perform_continuous_analysis){
+		  fit <- natmed2(
+		    W = data_full[, covariates, drop = FALSE],
+		    A = data_full$Trt,
+		    R = data_full$TwophasesampInd,
+		    S = data_full[, marker, drop = TRUE],
+		    C = data_full$Delta,
+		    Y = data_full$outcome,
+		    gRn = 1 / data_full$wt,
+		    glm_gA = "1",
+		    glm_gAS = NULL,
+	      	    SL_gAS = sl_library,
+	            glm_QY_WAS = NULL,
+	            SL_QY_WAS = sl_library,
+	            glm_QY_WACY = NULL,
+	            SL_QY_WACY = sl_library,
+	            glm_QD_WACY = NULL,
+	            SL_QD_WACY = sl_library,
+	            glm_QY_W = NULL,
+	            SL_QY_W = sl_library,
+	            glm_QY_WA = NULL,
+	            SL_QY_WA = sl_library
+		  )
+		  this_row <- format_row(fit)
+		}else{
+			this_row <- c(NA, NA, NA)
+		}
+		quant_result <- rbind(quant_result, this_row)
+
+		# categorical marker
+	  data_cat <- readRDS(here("data_clean", paste0("data_", time, "_cat.rds")))
+
+	  fit_cat <- natmed2(
+		    W = data_cat[, covariates, drop = FALSE],
+		    A = data_cat$Trt,
+		    R = data_cat$TwophasesampInd,
+		    S = data_cat[[paste0(marker, "cat")]],
+		    C = data_cat$Delta,
+		    Y = data_cat$outcome,
+		    gRn = 1 / data_cat$wt,
+		    glm_gA = "1",
+		    glm_gAS = NULL,
+	      	    SL_gAS = sl_library,
+	            glm_QY_WAS = NULL,
+	            SL_QY_WAS = sl_library,
+	            glm_QY_WACY = NULL,
+	            SL_QY_WACY = sl_library,
+	            glm_QD_WACY = NULL,
+	            SL_QD_WACY = sl_library,
+	            glm_QY_W = NULL,
+	            SL_QY_W = sl_library,
+	            glm_QY_WA = NULL,
+	            SL_QY_WA = sl_library
+		  )
+		this_row <- format_row(fit_cat)
+		cat_result <- rbind(cat_result, this_row)
+		day_col <- c(day_col, ifelse(grepl("Day57", marker), "Day 57", "Day 29"))
+		which_assay <- substr(marker, 6, nchar(marker))
+		assay_col <- c(assay_col, gsub("\\%", "\\\\\\%", labels.assays[which_assay]))
 	}
-	quant_result <- rbind(quant_result, this_row)
 
-	# categorical marker
-  data_cat <- readRDS(here("data_clean", paste0("data_", time, "_cat.rds")))
+	full_result <- cbind(day_col, assay_col, quant_result)
+	row.names(full_result) <- NULL
+	colnames(full_result) <- c("Time", "Assay", "Direct VE", "Indirect VE", "Prop. mediated")
+	full_result_cat <- cbind(day_col, assay_col, cat_result)
+	row.names(full_result_cat) <- NULL
+	colnames(full_result_cat) <- c("Time", "Assay", "Direct VE", "Indirect VE", "Prop. mediated")
 
-  fit_cat <- natmed2(
-	    W = data_cat[, covariates, drop = FALSE],
-	    A = data_cat$Trt,
-	    R = data_cat$TwophasesampInd,
-	    S = data_cat[[paste0(marker, "cat")]],
-	    C = data_cat$Delta,
-	    Y = data_cat$outcome,
-	    gRn = 1 / data_cat$wt,
-	    glm_gA = "1",
-	    glm_gAS = NULL,
-      	    SL_gAS = sl_library,
-            glm_QY_WAS = NULL,
-            SL_QY_WAS = sl_library,
-            glm_QY_WACY = NULL,
-            SL_QY_WACY = sl_library,
-            glm_QD_WACY = NULL,
-            SL_QD_WACY = sl_library,
-            glm_QY_W = NULL,
-            SL_QY_W = sl_library,
-            glm_QY_WA = NULL,
-            SL_QY_WA = sl_library
-	  )
-	this_row <- format_row(fit_cat)
-	cat_result <- rbind(cat_result, this_row)
-	day_col <- c(day_col, ifelse(grepl("Day57", marker), "Day 57", "Day 29"))
-	which_assay <- substr(marker, 6, nchar(marker))
-	assay_col <- c(assay_col, gsub("\\%", "\\\\\\%", labels.assays[which_assay]))
+	saveRDS(full_result, file = here("output", "full_result.rds"))
+	saveRDS(full_result_cat, file = here("output", "full_result_cat.rds"))
 }
-
-full_result <- cbind(day_col, assay_col, quant_result)
-row.names(full_result) <- NULL
-colnames(full_result) <- c("Time", "Assay", "Direct VE", "Indirect VE", "Prop. mediated")
-full_result_cat <- cbind(day_col, assay_col, cat_result)
-row.names(full_result_cat) <- NULL
-colnames(full_result_cat) <- c("Time", "Assay", "Direct VE", "Indirect VE", "Prop. mediated")
-
-saveRDS(full_result, file = here("output", "full_result.rds"))
-saveRDS(full_result_cat, file = here("output", "full_result_cat.rds"))

--- a/cop_mediation/report.Rmd
+++ b/cop_mediation/report.Rmd
@@ -7,11 +7,14 @@ wd <- here()
 root_dir <- gsub("/cop_mediation*", "", wd)
 renv::activate(root_dir)
 source(paste0(root_dir, "/_common.R"))
-full_result <- readRDS(paste0(root_dir, "/cop_mediation/output/full_result.rds"))
-full_result_cat <- readRDS(paste0(root_dir, "/cop_mediation/output/full_result_cat.rds"))
+analysis_ran <- file.exists(paste0(root_dir, "/cop_mediation/output/full_result.rds"))
+if(analysis_ran){
+	full_result <- readRDS(paste0(root_dir, "/cop_mediation/output/full_result.rds"))
+	full_result_cat <- readRDS(paste0(root_dir, "/cop_mediation/output/full_result_cat.rds"))
+}
 ```
 
-```{r, mediation-kable-quantitative, output = "asis"}
+```{r, mediation-kable-quantitative, output = "asis", eval = analysis_ran}
 init_kable <- knitr::kable(full_result, 
             format = "latex", row.names = FALSE, booktabs = TRUE, escape = FALSE,
             caption = paste0("Table of mediation effect estimates for quantitative markers with 95\\% confidence intervals. \\newline{} Direct VE = VE comparing vaccine vs. placebo with marker set to distribution in placebo. \\newline{} Indirect VE = VE in vaccinated comparing observed marker vs. hypothetical marker under placebo. \\newline{} Prop. mediated = fraction of total risk reduction from vaccine attributed to antibody response."),
@@ -22,10 +25,12 @@ if(any(is.na(full_result))){
 kableExtra::kable_styling(init_kable, latex_options = c("scale_down", "hold_position"))
 ```
 
-```{r, mediation-kable-categorical, output = "asis"}
+```{r, mediation-kable-categorical, output = "asis", eval = analysis_ran}
 init_kable_cat <- knitr::kable(full_result_cat, 
             format = "latex", row.names = FALSE, booktabs = TRUE, escape = FALSE,
             caption = paste0("Table of mediation effect estimates for tertile markers with 95\\% confidence intervals. \\newline{} Direct VE = VE comparing vaccine vs. placebo with marker set to distribution in placebo. \\newline{} Indirect VE = VE in vaccinated comparing observed marker vs. hypothetical marker under placebo. \\newline{} Prop. mediated = fraction of total risk reduction from vaccine attributed to antibody response."),
             linesep = c(rep("", length(assays)-1), "\\hline"))
 kableExtra::kable_styling(init_kable_cat, latex_options = c("scale_down", "hold_position"))
 ```
+
+`r if(!analysis_ran) "No mediation analysis was performed for these data."`

